Add request assertion feature to server

aquamme opened the pull request:
Implements #845 

The server tracks all handled requests and provides an API to allow
users to check whether a request matching the given parameters has
been made since the server was started.

``` javascript
  server.received.get.to('/contacts', { name: 'Darth Vader' })
  server.didNotReceive.post.to('any');
  server.received.get.to(/\/secrets.*/);
  server.didNotReceive.request.to('/robot/5');
  server.received.put.with({ name: 'Jack' });
```

I ended up adding RegExp support mostly because it was simpler than adding support for paths like "/posts/:id".

There is also a getter for the last request made to the server:

``` javascript
  server.lastRequest
```

Comments and suggestions are welcome!

aquamme commented on addon/request-assertion.js:
```diff
@@ -0,0 +1,90 @@
+/* * RequestAssertion implements a simple DSL for asserting that specific requests were made.
```
Not sure if `RequestAssertion` is the best name since 'assertion' is a loaded term. Maybe something like `RequestChecker`?

aquamme commented on addon/request-assertion.js:
```diff
@@ -0,0 +1,90 @@
+      Object.defineProperty(this, verb, {
+        get() {
+          let f = to.bind(null, requests, received, verb);
+          return { to: f, with: f };
```
`with` is an alias for `to`, only because it seems to read better in cases like this `server.received.put.with({ name: 'Todd' })` -- could remove it though.

aquamme commented:
A few other things I'd like to do/consider/I missed:
* Add an option to cap the number of requests that the server holds on to
* Add support for comparing POST payloads with nested attributes
* Have the feature report the number of matching requests instead of a boolean -- might want to rework the DSL a bit if this idea is pursued -- `server.received.post.to('/posts')` doesn't look like it would return a number

samselikoff commented:
Sorry about delay, hoping to get to this soon.

Just glancing over, will folks want to assert against content type or other things? Should we

```
assert.ok(this.server.received.post.to('/contacts', {
  with: {
    body: { name: 'Bilbo', age: 111 },
    headers: {
      'ContentType': 'foo'
    }
  })
);
```

Don't have to change right now, just a quick thought.

aquamme commented:
I like the idea. I'm on vacation until the 4th, so I won't be able to update the PR until then. Glad to hear any other feedback you may have when you get to it. Thanks!

samselikoff commented:
Few suggestions:

- Make `_requests` public: `server.requests`. This way folks can inspect during dev, or even use w/their assertions (`assert.equal(server.requests.length, 3);`)
- Move `server.lastRequest` to `server.requests.last` (I think you can define a property on the array)
- I think the DSL currently doesn't allow `server.received.post`, right? You have to specify a url? ... thinking about this, I think it's okay.
- Originally I said enable only in testing by default, but I think we should enable in all envs by default, and maybe add an option to disable if memory gets out of hand. (We can do this in a new PR).

aquamme commented:
Great feedback -- I should have time to make those changes tomorrow evening.

aquamme commented:
All right, this is ready for another peek.

`server.received...` returns the number of matching requests, while `server.didNotReceive` returns a boolean. So you can do things like
```javascript
// Two GETs are made to '/contacts', then...
assert.equal(server.received.get.to('/contacts'), 2);
assert.ok(server.received.get.to('/contacts'));
assert.ok(server.didNotReceive.post.to('/contacts'));
```
You can assert against any FakeXHR attributes:
```javascript
assert.ok(server.received.get.to('/contacts', { requestHeaders: { 'Foo': 'Bar' } }));
```

Also made `requests` a public property and added a `last` property.

> I think the DSL currently doesn't allow server.received.post, right? You have to specify a url? ... thinking about this, I think it's okay.

I don't see how we could make that syntax work while still allowing further chaining with 'to'. That aside, I did have a mechanism for allowing any URLs but I removed it because I didn't like the API (`server.received.request.to('any')`), but it could easily be added back in some form.

aquamme commented:
@samselikoff -- any further changes you'd like to see on this? I can write up some documentation once you give it the 👍

samselikoff commented:
Closing this as stale, but keeping a reference to it in https://github.com/miragejs/miragejs/issues/217 in case we want to reference these APIs.
